### DISCUSSION TITLE
Improve error messages when password doesn't meet policy requirements.

### DISCUSSION THREAD

**Discussion by TonyLovesDevOps** [directus/directus#12788] [Feature Requests]
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues) that already describes my problem.

### Describe the Bug

Right now, when a user offers a password that doesn't meet the password complexity requirements, the error messages returned to the user could be more helpful.

For example, when accepting an invite, the user sees an `Unexpected Error`:
<img width="427" alt="Unexpected error" src="https://user-images.githubusercontent.com/868644/136590659-df046cda-3f60-4a2c-89ec-8f71d12ad479.png">

And when resetting a password, they receive the much better but still not super clear `Value doesn't have the correct format`:
<img width="423" alt="Value doesn't have the correct format " src="https://user-images.githubusercontent.com/868644/136591131-b26ec800-c99c-484f-bc16-0e671288e876.png">

### To Reproduce

Scenario 1: Accept invite
1. Choose an `Auth Password Policy` other than `None` on the `/admin/settings/project` page;
2. Send an email invite to a new user;
3. Click the invite link and enter a password that doesn't meet the requirements;

Scenario 2: Reset password
1. Choose an `Auth Password Policy` other than `None` on the `/admin/settings/project` page;
2. Log in as an existing user and attempt to reset their password to a new one that doesn't meet the policy requirements.

Expected: User receives an error message stating that their password does not meet the password policy.
Actual: User receives `Unexpected Error` in scenario 1, and `Value doesn't have the correct format` in scenario 2.

In both cases, the directus logs contain messages showing `Provided password doesn't match password policy` -- can we return that to the user in both cases instead of the cryptic messages?

### What version of Directus are you using?

v9.0.0-rc.96

### What version of Node.js are you using?

v16.10.0 (from directus/directus:9.0.0-rc.96 docker image)

### What database are you using?

MariaDB 10.3.23

### What browser are you using?

Chrome 94.0.4606.71

### What operating system are you using?

macOS

### How are you deploying Directus?

Docker

**Comment by TonyLovesDevOps** [comment]
One idea around this would be to add an element to the `Auth Password Policy` section with an optional `Custom error message` or `Description of policy` field that could be shown to the users in this case, in addition to or instead of the `Provided password doesn't match password policy` message, e.g. `Password must be a minimum of 20 characters in length with at least one number, one uppercase letter, and one symbol`.

**Comment by azrikahar** [comment]
This was actually my concern over at https://github.com/directus/directus/pull/8526#issuecomment-933290682 🙈 

> In both cases, the directus logs contain messages showing Provided password doesn't match password policy -- can we return that to the user in both cases instead of the cryptic messages?

We might need to change the exception thrown here:

https://github.com/directus/directus/blob/969621249fbcffe46d2106c44ad53c05a07d8d2c/api/src/services/users.ts#L92-L99

so that we can translate the message on the app 🤔

> One idea around this would be to add an element to the `Auth Password Policy` section with an optional `Custom error message` or `Description of policy` field that could be shown to the users in this case,

Interesting idea!

**Comment by joselcvarela** [comment]
Should this be moved into Ideas in discussions?
Does not seem a bug to me, just an improvement.

**Comment by rijkvanzanten** [comment]
@joselcvarela I was doubting that.. I didn't want to call it a "bug" as nothing is technically broken or not working as expected, _however_ it's a little silly for that page to say "Unexpected error", which could be considered a "bug" in the UX, so I left it up 👍🏻

**Comment by rijkvanzanten** [comment]
@azrikahar Adding a new "PasswordPolicyValidationException" would make the most sense to me. As for the translated message, I agree we can't just show "Password should match `/(?=^.{8,}$)(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[!@#$%^&*()_+}{';'?>.<,])(?!.*\s).*$/`", as nobody would understand what that means.. Just showing "Password isn't strong enough" wouldn't really help much neither. 

Adding a custom message would be _nice_, but would still require you to explain the regex you entered again. Technically speaking, we have all the information we need in the regex itself:

`/(?=^.{8,}$)(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[!@#$%^&*()_+}{';'?>.<,])(?!.*\s).*$/`

| Part | Description | 
|-----|-------------|
| `(?=^.{8,}$)` | At least 8 characters |
| `(?=.*\d)` | Needs to have a number |
| `(?=.*[a-z])` | Needs to have a lowercase character |
| `(?=.*[A-Z])` | Needs to have an uppercase character |
| `(?=.*[!@#$%^&*()_+}{';'?>.<,])` | Needs to have one of `!@#$%^&*()_+}{';'?>.<,` |
| `(?!.*\s)` | Can't have any spaces |

I'm wondering if we could pull a regex101.com, and basically come up with a human readable description of the given regex 🤔

**Reply by inta** [reply]
Whats the idea of exluding spaces? I don't think its a good idea to limit the entropy in any way.

**Reply by rijkvanzanten** [reply]
This is just an example of how we could potentially check parts in a regex

**Reply by inta** [reply]
Really? I found this discussion because I found out Directus is refusing every passwort with a space in it if the rule is set to strong. That was unexpected for me.

**Reply by paescuj** [reply]
> Really? I found this discussion because I found out Directus is refusing every passwort with a space in it if the rule is set to strong. That was unexpected for me.

Well, the RegEx from above indeed corresponds to the one from the "strong" policy in Directus and I do think it's arguable whether spaces should be allowed, too.
However, the discussion here was about a different topic, so it would be better to open up a separate one for this concern.

**Comment by TonyLovesDevOps** [comment]
> I'm wondering if we could pull a regex101.com, and basically come up with a human readable description of the given regex 🤔

🧐 regex101.com does have a [public API](https://github.com/firasdib/Regex101/wiki/API#create-an-entry-) that could even be used to create a regex and then point the user to a URL that would include the `Explanation` -- though the verbosity of the explanation is probably almost as impenetrable as a regex to a normal user (not to mention creating a dependency on a third-party service with zero availability guarantees). I took a quick look for libraries that turn a regex into something human readable and didn't see anything promising at first glance, and whatever regex101 uses is not open source.

This is where I thought a custom error message might come in handy... if you're the admin setting the password policy, you're going to want to communicate it to your users _somehow_, unless you're a BOFH in which case `Unexpected Error` seems a perfect way to punish your (l)users. 😈

**Comment by rijkvanzanten** [comment]
> 🧐 regex101.com does have a public API that could even be used to create a regex and then point the user to a URL that would include the Explanation -- though the verbosity of the explanation is probably almost as impenetrable as a regex to a normal user (not to mention creating a dependency on a third-party service with zero availability guarantees).

Yeah I wouldn't want to actually point to regex101, I meant that it would be nice if we could do a similar trick where we auto-generate a human readable explanation 🙂

**Comment by azrikahar** [comment]
> Adding a new "PasswordPolicyValidationException" would make the most sense to me.

@rijkvanzanten Yea that's what I had in mind as well. Might be the "MVP" approach in this case.

> I'm wondering if we could pull a regex101.com, and basically come up with a human readable description of the given regex 

That's super interesting! But it is uncharted territory for me 😄 Maybe we can somehow do something with the AST we get after using [regexp-tree](https://www.npmjs.com/package/regexp-tree)? 

I've also stumbled upon this site: https://regexper.com/. The railroad diagram approach somewhat make things clearer from a technical perspective, but I also understand that this is no where near understandable for non-technical end users.

**Comment by joselcvarela** [comment]
I keep questioning myself why do websites require to have special characters, uppercase, lowercase, numbers for passwords.
If the image below tells the truth we could only ask user to put a long password.
For example, I have some accounts on Google using a long string all lowercase.
On this cases, we could suggest users to put a sentence from a lyrics they love 😀

<img width="400" src="https://user-images.githubusercontent.com/14039341/137065887-7135cb0e-6c09-46fc-b004-9cf734960c71.png" />

About the regex, maybe we could use named groups somehow: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Groups_and_Ranges#using_named_groups

**Comment by paescuj** [comment]
Might be interesting as well: https://github.com/VerbalExpressions/JSVerbalExpressions

**Comment by benhaynes** [comment]
All good stuff! I'd like to add the recommendation from @TonyLovesDevOps — let's add a `Password Policy Error Message` field (STRING) that allows entering a custom message. The default value here should be:

```
Doesn't fulfill password policy requirements
```

The input should be full width, and below the other two fields.

**Comment by paescuj** [comment]
> All good stuff! I'd like to add the recommendation from @TonyLovesDevOps — let's add a `Password Policy Error Message` field (STRING) that allows entering a custom message. The default value here should be:
> 
> ```
> Doesn't fulfill password policy requirements
> ```
> 
> The input should be full width, and below the other two fields.

While a message like this is way better than `Unexcepted Error`, it still isn't very helpful for users as they don't know how to full-fill the requirements.
It would be awesome to have an interface to define policies which are then automatically turned into human readable messages 😄
I also stumbled upon the follwing libraries which might be helpful:
* https://github.com/auth0/password-sheriff
* https://github.com/kamronbatman/joi-password-complexity

**Comment by benhaynes** [comment]
> While a message like this is way better than Unexcepted Error, it still isn't very helpful for users as they don't know how to full-fill the requirements.

Why doesn't this direction solve for that? The placeholder value is only because it's not dynamic... but the admin would change it to a specific that matches the Policy/RegEx. So if they make some crazy policy, they can say:

```
Must contain 3 multi-case letters, 4 numbers, 6 special characters, and an emoji.
```

To me, this adds the most freedom and avoids us building/integrating/maintaining a complex regex decoder.

**Comment by paescuj** [comment]
> > While a message like this is way better than Unexcepted Error, it still isn't very helpful for users as they don't know how to full-fill the requirements.
> 
> Why doesn't this direction solve for that? The placeholder value is only because it's not dynamic... but the admin would change it to a specific that matches the Policy/RegEx. So if they make some crazy policy, they can say:
> 
> ```
> Must contain 3 multi-case letters, 4 numbers, 6 special characters, and an emoji.
> ```
> 
> To me, this adds the most freedom and avoids us building/integrating/maintaining a complex regex decoder.

True, should have read @TonyLovesDevOps proposal more carefully 🙈👍

**Comment by rijkvanzanten** [comment]
One thing to keep in mind with the custom error message is i18n... 😬

**Comment by benhaynes** [comment]
True, but that will be a bigger task with the "string" approach... as we don't support it for the "Public Note" either.

**Comment by aidenfoxx** [comment]
You could also just break down the components of password complexity into configurable options in the UI. Realistically that should serve 90% of users, and you have fixed parameters to return a message by. E.g:

- Minimum length input
- Require numeric checkbox
- Require uppercase checkbox
- Require symbol checkbox

That would also allow for a nicer error interface à la:

![image](https://user-images.githubusercontent.com/5548518/137167011-22411fbb-13ab-428c-81c9-449278d8ceec.png)

**Comment by TonyLovesDevOps** [comment]
@aidenfoxx @rijkvanzanten @paescuj do any of you have code brewing for this, or an idea of when you might spend time on it? Alternatively, I could take a stab at the `Password Policy Error Message` implementation, which isn't as fancy as some of the other cool ideas posted here, but would definitely work for my purposes.

**Comment by paescuj** [comment]
@TonyLovesDevOps Not from my side... Would love to see a PR from your side 😃❤️‍🔥

**Comment by rijkvanzanten** [comment]
Likewise, I don't have anything in works at the moment 👍🏻

**Comment by aidenfoxx** [comment]
Nothing from me. Go for it! 👍 😄

**Comment by TonyLovesDevOps** [comment]
Cool, I opened a PR for this.  I'd value reviews/input on the PR, especially if someone knows of a boilerplate unit test I can copy as a starting point for this - I didn't see any existing tests that set up an `auth_password_policy` scenario, and I don't have time right now to bang my head against an unfamiliar unit testing framework. 😭 

Here's what it looks like with a custom message when accepting an invite / changing a password:

<img width="400" alt="accept-invite-validation-error" src="https://user-images.githubusercontent.com/868644/137971920-1a6978b2-7b97-4e8c-bb30-73b0074935f8.png">

<img width="538" alt="change-password-validation-error" src="https://user-images.githubusercontent.com/868644/137971608-3d7abd5d-c1c5-466a-8fe0-d8396d14c46b.png">

**Comment by TonyLovesDevOps** [comment]
So #8946 appears to be in PR purgatory. Is there anything I can do to help get it merged?

I'd happily support a better alternative implementation but we'll soon need to consider the regrettable decision to fork so that our users have a better experience when accepting an email invite, as almost no one meets our policy requirements on the first try.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Reply by northern-mindset** [reply]
In this case I'd consider this to be a bug report rather than a feature request…

**Comment by abernh** [comment]
For everybody stuck with the ambiguous error message (e.g. on Directus Cloud without options to fix it) I posted a CSS-only hack here 

https://github.com/directus/directus/discussions/13896#discussioncomment-6275872

**Comment by thorstenvonberswordt** [comment]
This all is an absolutely valid complaint (just as the: wrong payload message on double password creation -> discussion), and should be fixed since it seems reasonably simpel to do so! I too stumbled quite fast across this problem in our evaluation of directus. And for sure if you want to market this as headless cms those kind of things got to be in order since not everybody wants to program an API interface just to be able to add users ....
love the css hack from @abernh but this should be necessary!
